designs/simplified-scaffolding.md

```diff
@@ -0,0 +1,512 @@
+If we want to help users test that their object's structure is the way
+they think it is, we're probably better served coming up with a standard
+"can I create this example YAML file".
+
+Furthermore, since the structure is quite convoluted, it makes it more
+difficult to write examples, since the actual code we care about ends up
+scattered deep in a folder structure.
+
+### Lack of Builder
+
+We introduced the builder pattern for controller construction in
+controller-runtime
+([GoDoc](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/builder#ControllerManagedBy))
+as a way to simplify construction of controllers and reduce boilerplate
+for the common cases of controller construction.  Informal feedback from
+this has been positive, and it enables fairly rapid, clear, and consice
+construction of controllers (e.g. this [one file
+controller](https://github.com/DirectXMan12/sample-controller/blob/workshop/main.go)
+used as a getting started example for a workshop).
+
+Current KubeBuilder scaffolding does not take advantage of the builder,
+leaving generated code using the lower-level constructs which require more
+understanding of the internals of controller-runtime to comprehend.
+
+### Dependency Passing Woes
+
+Another common line of questioning amongst KubeBuilder users is "how to
+I pass dependencies to my controllers?".  This ranges from "how to I pass
+custom clients for the software I'm running" to "how to I pass
+configuration from files and flags down to my controllers" (e.g.
+[kubernete-sigs/kubebuilder#611](https://github.com/kubernetes-sigs/kubebuilder/issues/611)
+
+Since reconciler implementations are initialized in `Add` methods with
+standard signatures, dependencies cannot be passed directly to
+reconcilers.  This has lead to requests for dependency injection in
+controller-runtime (e.g.
+[kubernetes-sigs/controller-runtime#102](https://github.com/kubernetes-sigs/controller-runtime/issues/102)),
+but in most cases, a structure more amicable to passing in the
+dependencies directly would solve the issue (as noted in
+[kubernetes-sigs/controller-runtime#182](https://github.com/kubernetes-sigs/controller-runtime/pull/182#issuecomment-442615175)).
+
+## Revised Structure
+
+In the revised structure, we use the builder pattern to focus on the
+"code-refactor-code-refactor" cycle: start out with a simple structure,
+refactor out as your project becomes more complicated.
+
+Users receive a simply scaffolded structure to start. Simple projects can
+remain relatively simple, and complicated projects can decide to adopt
+a different structure as they grow.
+
+The new scaffoled project structure looks something like this (compiled
+artifacts like config omitted for brevity):
+
+```shell
+$ tree ./test/project
+./test/project
+├── main.go
+├── controller
+│   ├── mykind_controller.go
+│   ├── mykind_controller_test.go
+│   └── controllers_suite_test.go
+├── api
+│   └── v1
```

Comment:
It sounds like this is the proposed structure if we have more than one versions.
```
├── api
│   ├── v1
│   │   └── foo_types.go
│   │   └── groupversion_info.go
│   └── v2
│       └── foo_types.go
│       └── groupversion_info.go
```
It sounds like the info about group will be duplicated across versions.
Given we assume only one group in this structure, does it make sense to have structure like
```
├── api
│   ├── group_info.go
│   ├── v1
│   │   └── foo_types.go
│   │   └── version_info.go
│   └── v2
│       └── foo_types.go
│       └── version_info.go
```

Reply:
good call